src/NServiceBus.Core.Tests/Pipeline/EnsureNativeBehaviors.cs

```diff
@@ -0,0 +1,30 @@
+﻿namespace NServiceBus.Core.Tests.Pipeline
+{
+    using System.Linq;
+    using NServiceBus.Pipeline;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class EnsureNativeBehaviors
+    {
+        [Test]
+        public void CoreBehaviorsMustNotUseAbstractClass()
+        {
+            var behaviorTypes = typeof(IBehavior).Assembly.GetTypes()
+                .Where(type => typeof(IBehavior).IsAssignableFrom(type))
+                .ToArray();
+
+            foreach (var type in behaviorTypes)
+            {
+                for (var baseType = type.BaseType; baseType != null && baseType != typeof(object); baseType = baseType.BaseType)
+                {
+                    if (baseType.IsGenericType)
+                    {
+                        var genericType = baseType.GetGenericTypeDefinition();
+                        Assert.IsFalse(genericType == typeof(Behavior<>), $"For performance reasons, built-in behavior `{type}` is not allowed to inherit from abstract class Behavior<T>. Implement IBehavior<Tin, TOut> directly, using the same context type for both TIn and TOut.");
+                    }
+                }
+            }
+        }
+    }
```

Comment:
Another possible approach, based on what I did in https://github.com/Particular/NServiceBus/pull/5850:

```suggestion    using System;
    using System.Linq;
    using NServiceBus.Pipeline;
    using NUnit.Framework;

    [TestFixture]
    public static class EnsureNativeBehvaiors
    {
        [Test]
        public static void CoreBehaviorsMustNotUseAbstractClass()
        {
            var violators = typeof(IBehavior).Assembly.GetTypes()
                .Where(UsesAbstractClass)
                .ToList();

            Console.Error.WriteLine($"Violators of {nameof(CoreBehaviorsMustNotUseAbstractClass)}:{Environment.NewLine}{string.Join(Environment.NewLine, violators)}");

            Assert.IsEmpty(violators, $"For performance reasons, built-in behaviors are not allowed to inherit from abstract class Behavior<T>. Implement IBehavior<Tin, TOut> directly, using the same context type for both TIn and TOut.");
        }

        static bool UsesAbstractClass(this Type type) =>
            type != null && type != typeof(Behavior<>) && ((type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Behavior<>)) || type.BaseType.UsesAbstractClass());
    }
```